zebrad/src/application.rs

```diff
@@ -135,10 +136,28 @@ impl Application for ZebradApp {
 
 impl ZebradApp {
     fn level(&self, command: &EntryPoint<ZebradCmd>) -> String {
-        if let Ok(level) = std::env::var("ZEBRAD_LOG") {
-            level
-        } else if command.verbose {
+        use ZebradCmd::*;
+        // Tracing sends output to stdout, so we disable it for some commands.
+        //
+        // TODO: send tracing output to stderr. This change requires an abscissa
+        //       update, because `abscissa_core::component::Tracing` uses
+        //       `tracing_subscriber::fmt::Formatter`, which has `Stdout` as a
+        //       type parameter. We need `MakeWriter` or a similar type.
+        let command_uses_stdout = match command.command {
+            Some(Generate(_)) | Some(Help(_)) | Some(Revhex(_)) | Some(Version(_)) => true,
+            _ => false,
+        };
+
+        // Allow users to:
+        //  - override all other configs and defaults using the command line
+        //  - see command outputs without spurious log messages, by default
+        //  - override the config file using an environmental variable
+        if command.verbose {
             "debug".to_string()
+        } else if command_uses_stdout {
+            "warn".to_string()
+        } else if let Ok(level) = std::env::var("ZEBRAD_LOG") {
```

Comment:
I think the env variable should take priority over the `command_uses_stdout` default

Reply:
unrelated but, wait, is this why `RUST_LOG` doesn't affect Zebra settings?

Reply:
yea, I thought it might be a good idea, I hate when I mess with `RUST_LOG` and my terminal prompt / cargo both start outputting log lines.

Reply:
Ah okay, makes sense! It's news to me. We should probably document this.

Reply:
> I think the env variable should take priority over the `command_uses_stdout` default

What's the use case for logging during Generate, Help, Revhex, or Version?

I chose to ignore the env var, because people usually set env vars in their shell profile, and then forget about them. It would be surprising to me to get log output in my generated config, even if I set an env var.

Reply:
Whatever we decide, I should modify the zebrad usage to document logging via the env var and verbose flag.

Reply:
- [ ] I should also turn off logging for the no-argument case, which just prints usage. Unless we're going to make that default to `start`?

Reply:
> It would be surprising to me to get log output in my generated config, even if I set an env var.

It also makes the generated config unusable.

I'm happy to revise the order once abscissa is fixed.

Reply:
> What's the use case for logging during Generate, Help, Revhex, or Version?
> 
> I chose to ignore the env var, because people usually set env vars in their shell profile, and then forget about them. It would be surprising to me to get log output in my generated config, even if I set an env var.

This seems to be a difference in assumption, for me I'd assume people wouldn't ever permanently export a log level changing variable, and that each time it gets specified it would be intentional. So my assumption is that most of the time people still wouldn't be logging during Generate, Help, etc. I don't really mind picking to de-prioritize the env variable, this will all be a moot point once we fix the issue in abscissa anyways.

Reply:
>     * [ ]  I should also turn off logging for the no-argument case, which just prints usage. Unless we're going to make that default to `start`?

we're not gonna make `start` the default, so go ahead and disable logging for this case IMO

Reply:
I think it's reasonable to suppress all logging output for these commands (like @teor2345 is doing now).

Reply:
>     * Whatever we decide, I should modify the zebrad usage to document logging via the env var and verbose flag.

I couldn't work out how to do this quickly, so I opened #656 to do it eventually.

Reply:
>  I should also turn off logging for the no-argument case, which just prints usage. Unless we're going to make that default to `start`?

Fixed in a9c03a8b8e3d8458c679e08f2ff935f7ed8d2436